Better algorithm for optimization in 'crush-compat' mode

SpandanKumarSahu opened the pull request:
Utilize the optimized algorithm used in Loic's python-crush to set the new weights in crush-compat mode.

Justifications for this algorithm:
1. In the existing algorithm, we change the weights of the OSDs, in such a manner that the total sum of the weights of all OSDs also changes. This makes it difficult to maintain/achieve the desired "relative weight" for a particular OSD.
2. The assigning of new_weight as  0.7 * old_weight + 0.3 * new_weight, prevents faster rate of optimization. (For e.g, if the new_weight was in fact the true optimized weight, then according to the current algorithm, it would theoretically have never attained that value)

liewegas committed:
mgr: add trivial OSDMap wrapper class

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/PyModules: add 'pg_status' dump

This is summary info, same as what's in 'ceph status'.

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
pybind/mgr/mgr_module: add default arg to get_config

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
pybind/mgr/balancer: add balancer module

- wake up every minute
- back off when unknown, inactive, degraded
- throttle against misplaced ratio
- apply some optimization step
  - initially implement 'upmap' only

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
pybind/mgr/balancer: do upmap by pool, in random order

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
crush/CrushWrapper: refactor get_rule_weight_osd_map to work with roots too

Allow us to specify a root node in the hierarchy instead of a rule.
This way we can use it in conjunction with find_takes().

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
crush/CrushWrapper: fix output arg for find_{takes,roots}()

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/PyOSDMap: get_crush, find_takes, get_take_weight_osd_map

These let us identify distinct CRUSH hierarchies that rules distribute
data over, and create relative weight maps for the OSDs they map to.

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/PyModules: add 'pg_dump' get

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
crush/CrushWrapper: rule_has_take

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/PyOSDMap: OSDMap.map_pool_pgs_up, CRUSHMap.get_item_name

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
pybind/mgr/balancer: rough framework

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
...

liewegas committed:
pybind/mgr/balancer: make 'crush-compat' sort of work

Signed-off-by: Sage Weil <sage@redhat.com>

SpandanKumarSahu committed:
src/pybind/mgr/balancer/module.py: improve scoring method

* score lies in [0, 1), 0 being perfect distribution
* use shifted and scaled cdf of normal distribution
  to prioritize highly over-weighted device.
* consider only over-weighted devices to calculate score

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu committed:
squash with b3f9a723b453c13868873a3be1d7c23b894843be

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu committed:
src/pybind/mgr/balancer/module.py: better do_crush_compat

Utilize the optimized algorithm used in Loic's python-crush to set the
new weights in crush-compat mode

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu committed:
src/pybind/mgr/balancer/module.py: add cli to set 'key'

`do_crush_compat` utilizes 'key' (which can have values as 'pgs', 'objects', 'bytes') to predict the new weights.
'key' should be set using command rather than hard-coding it.

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

liewegas committed:
Merge pull request #59 from SpandanKumarSahu/balancer_score

src/pybind/mgr/balancer/module.py: improve scoring method

liewegas commented:
I like the idea of moving weight around to keep the totals sane.  But it seems like this whole section should be in a loop so we can optimize.  i.e., after each modification (or set of modifications), do ms = plan.final_state() and calc_eval() that to see what the result is?

SpandanKumarSahu commented:
Yes, that is what was being done in Loic's work too. The step of optimization was done iteratively for a fixed number of times (or till convergence was achieved, whichever was smaller).
I first wanted to figure out a simple implementation and then port other things from the python-crush.

liewegas commented:
I wonder if it's easier to just adjust the weights without thinking too much about the total, and then renormalize at the end?  I think we'll need a renormalize method anyway because we might remove an item that is below or above it's original weight and that will throw off the weight-set total one way or the other.

The reason why i did hte .7 thing is because when i jumped straight to what it looked like it should be we didn't converge.. it oscillated around the real value.  In practice I think we want to iteratively step toward the optimal value.  Maybe the best next step is to restructure this function so that it tests its changes in a loop before committing to a particular set of weight changes...?

liewegas committed:
pybind/mgr/balancer: make auto mode work

(with upmap at least)

Signed-off-by: Sage Weil <sage@redhat.com>

SpandanKumarSahu committed:
src/pybind/mgr/balancer/module.py: iterate simulation for better weights

* add convergence checking algorithm (Kullback–Leibler divergence test)
* iterate until converged or max_iterations

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu committed:
src/pybind/mgr/balancer/module.py: set cli for `max-iterations`

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu committed:
src/pybind/mgr/balancer/module.py: add 'auto' key for cursh-compat mode

* provide cli to set and view key-weights
* choose 'key' in crush-compat mode optimization, automatically, based on the relative weights of the keys

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu commented:
* Rebased
* Now the algorithm iterates until the convergence, or max_iterations (which can be set using cli) whichever is lower
* Added  'auto' as a key. When 'key' is set as 'auto', then optimization is done for all the keys (pgs, bytes and objects), depending upon the relative weights for the keys, set using 'key-weights' (implemented as a cli)

SpandanKumarSahu commented:
AFAIK, when we call `plan.final_state()` , the changes are actually committed. We should have a separate function (say, calculate_new_MappingState) at [here,](https://github.com/liewegas/ceph/pull/60/files#diff-aaf33d75795c314c9b5232fb40490fb7R675) that should calculate the new MappingState, without actually committing the new weights.

How should we go about implementing that function?

liewegas commented:
final_state() will apply teh changes to an in-memory OSDMap, but it's a private copy just for our own use (to pass to calc_eval etc).   No cluster changes are ever made except by execute() (which issues mon commands).

liewegas commented:
THis sounds great!  I'll take a closer look at it later today

tchaikov commented on src/pybind/mgr/balancer/module.py:
```diff
@@ -575,6 +608,20 @@ def do_upmap(self, plan):
+    def get_score(pe, key):
+        score = 0.0
+        if key == 'auto':
+            for r, vs in pe.score_by_root.iteritems():
```
use '_' for the unused key

tchaikov commented on src/pybind/mgr/balancer/module.py:
```diff
@@ -603,31 +650,118 @@ def do_crush_compat(self, plan):
+        best_weights = {}
+
+        key_copy = self.get_config('key', default_key)
+        if(key == 'auto'):
```
remove the parenthesies

tchaikov commented on src/pybind/mgr/balancer/module.py:
```diff
@@ -603,31 +650,118 @@ def do_crush_compat(self, plan):
-            for osd in queue:
-                deviation = target[osd] - actual[osd]
-                if deviation == 0:
+        for iterations in range(max_iterations):
```
xrange if py3 support is not expected.

SpandanKumarSahu replied:
In case of py2, `range()` will actually return a list instead of a generator. So, the only problem will be that we will need a bit more space in memory.
But if we keep it this way, the code will be support both py2 and py3, instead of having to write separate codes for py2 and py3 and also the additional need of importing `sys` module to check the python version.

tchaikov commented on src/pybind/mgr/balancer/module.py:
```diff
@@ -226,6 +242,9 @@ class Module(MgrModule):
     run = True
     plans = {}
     mode = ''
+    key = 'pgs'
```
where is this class variable referenced?

SpandanKumarSahu replied:
It has been referenced in 
`do_crush_compat` function at [here](https://github.com/liewegas/ceph/pull/60/files#diff-aaf33d75795c314c9b5232fb40490fb7R661).

SpandanKumarSahu committed:
doc/mgr: add balancer.rst
doc/man/8/ceph.rst: update ceph.rst to include balancer
src/pybind/mgr/balancer/module.py: improve documentation

provide basic description of the balancer module.

Signed-off-by: Spandan Kumar Sahu <spandankumarsahu@gmail.com>

SpandanKumarSahu commented:
Improved documentation.
Added balancer.rst in doc/mgr

tchaikov commented on src/pybind/mgr/balancer/module.py:
```diff
@@ -674,6 +674,10 @@ def do_crush_compat(self, plan):
         for iterations in range(max_iterations):
             ms = plan.final_state()
             pe = self.calc_eval(ms)
+            '''
```
please don't use literal string for commenting the code.

tchaikov commented on src/pybind/mgr/balancer/module.py:
```diff
@@ -685,6 +689,7 @@ def do_crush_compat(self, plan):
                 else:
                     key = 'bytes'
 
+            # Convergence Testing Algorithm
```
and please squash this change into the commit where the commented code is introduced.

SpandanKumarSahu replied:
There was another instance where I used literal string for commenting out the code, but the branch has been merged.
Should I change the commenting style there too?

tchaikov replied:
please add a commit to address it and add the SHA1 of the fixed commit in its commit message so Sage can help squash it before merging his branch into master.

tchaikov commented on doc/mgr/balancer.rst:
```diff
@@ -0,0 +1,188 @@
+OSDs. Even if the PGs, were balanced properly, with respect to the weights
+of the OSDs, there is still chance of uneven distribution, because the number
+of objects assigned to different PGs are not in required proportion. And even
+if they were, the sizes of each object might vary.
```
how to tell if the distribution in a cluster is uneven?

SpandanKumarSahu commented on doc/mgr/balancer.rst:
```diff
@@ -0,0 +1,188 @@
+While it is recommended that the balancer module be best left to run auto-magically,
+it might interest certain users to turn off the automatic optimization, and instead
+create optimization plans themselves, and execute them, whenever they feel the need
+to do so. While there is no fixed way to determine when to execute the rebalance
```
@tchaikov By this, we will be able to tell how uneven the distribution in the cluster is.

SpandanKumarSahu commented:
I have described the same in here
<https://github.com/liewegas/ceph/pull/60/commits/3d71345987ad8083785bac168e6b0092a3b2ebd4#diff-56b33f36ce15831e33d81fdf97404b43R23>
.

On Thu, Aug 24, 2017 at 3:45 PM, Kefu Chai <notifications@github.com> wrote:

> *@tchaikov* commented on this pull request.
> ------------------------------
>
> In doc/mgr/balancer.rst
> <https://github.com/liewegas/ceph/pull/60#discussion_r134978043>:
>
> > @@ -0,0 +1,188 @@
> +balancer plugin
> +================
> +
> +Typically, if you say, the system is 70% filled, then, all the OSDs,
> +should be 70% filled (roughly). It shouldn’t be the case that, a particular
> +OSD is 90% filled whereas another one is 20% filled. If such a case exist,
> +then we say there is an uneven distribution. Uneven distribution may arise
> +out of a number of reasons, including improper distribution of PGs among the
> +OSDs. Even if the PGs, were balanced properly, with respect to the weights
> +of the OSDs, there is still chance of uneven distribution, because the number
> +of objects assigned to different PGs are not in required proportion. And even
> +if they were, the sizes of each object might vary.
>
> how to tell if the distribution in a cluster is uneven?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/liewegas/ceph/pull/60#pullrequestreview-58334023>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQs9HR8OXFdvUhr0twBUIMrSBcrgWuZIks5sbU0ngaJpZM4Ov5YZ>
> .
>



-- 
Regards
Spandan Kumar Sahu

tchaikov commented on doc/mgr/balancer.rst:
```diff
@@ -0,0 +1,285 @@
+with 0 meaning perfect distribution, and 1 meaning the average over-utilized device
+is infinitely more weighted than the average utilization of all devices.
+
+As a thumb-rule, if score is 0.08, then the average over-utilized device is 10%
```
rule of thumb.

tchaikov commented on doc/mgr/balancer.rst:
```diff
@@ -0,0 +1,285 @@
+implies 100% more weighted. The scoring is non-linear, to signify that presence of
+highly over-utilizaed devices is a pretty bad situation.
+
+Wokring Process and Performance
```
Working.

tchaikov commented on doc/mgr/balancer.rst:
```diff
@@ -0,0 +1,285 @@
+At each instance of optimization, balancer module creates a `Plan`. Then it uses the
+CRUSH to identify the various OSDs in a pool and calculate the distribution of `pg`,
+`bytes` and `objects`, from the OSDMap. Then it calculates, how the target distribution
+of each of the parameter(`pg`, `bytes` and `objects`). Using an iterative gradient
```
is

> Using an iterative gradient  descent algorithm

a complete sentence or a part of it?

SpandanKumarSahu replied:
It was supposed to be a complete sentence, but I guess, it became a bit informal. Updated.

tchaikov commented on doc/mgr/balancer.rst:
```diff
@@ -0,0 +1,285 @@
+the weights get implemented. If the process converged, or the maximum number of iterations
+were made, then the best-weights during the entire process is recorded in the plan. The module
+then sends a `Mon` command, to actually execute the new-weights, and then remove the plan
+from the memory. If the balancer is set on autmagic mode, this entire process, gets repeated
```
`autmagic`?
